### PR TITLE
[mlir][Vector] Remove uses of vector.extractelement/vector.insertelement

### PR DESCRIPTION
**Pull request by Groverkss**
This patch removes usages of vector.extractelement/vector.insertelement. These operations can be fully represented by vector.extract/vector.insert. See https://discourse.llvm.org/t/rfc-psa-remove-vector-extractelement-and-vector-insertelement-ops-in-favor-of-vector-extract-and-vector-insert-ops/71116 for more information.

### CODE REVIEW

**Comment by Groverkss** [comment]
Depends on https://github.com/llvm/llvm-project/pull/113828

Please only review the top commit until that patch lands

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp [hunk 1]
```diff
@@ -2268,7 +2267,8 @@ LogicalResult mlir::linalg::vectorizeCopy(RewriterBase &rewriter,
       loc, readType, copyOp.getSource(), indices,
       rewriter.getMultiDimIdentityMap(srcType.getRank()));
   if (cast<VectorType>(readValue.getType()).getRank() == 0) {
-    readValue = rewriter.create<vector::ExtractElementOp>(loc, readValue);
+    readValue = rewriter.create<vector::ExtractOp>(loc, readValue,
+                                                   SmallVector<int64_t>{});
```
Comment (kuhar) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:2271]: You should be able to pass an empty array ref here?

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp [hunk 2]
```diff
@@ -1415,7 +1413,8 @@ vectorizeAsLinalgGeneric(RewriterBase &rewriter, VectorizationState &state,
     // 3.c. Not all ops support 0-d vectors, extract the scalar for now.
     // TODO: remove this.
     if (readType.getRank() == 0)
-      readValue = rewriter.create<vector::ExtractElementOp>(loc, readValue);
+      readValue = rewriter.create<vector::ExtractOp>(loc, readValue,
+                                                     SmallVector<int64_t>{});
```
Comment (kuhar) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1417]: You should be able to pass an empty array ref here?

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp (hunk 1 above)
Comment (kuhar) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:2271]: Also elsewhere in this PR

File: mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp [hunk 3]
```diff
@@ -756,7 +756,8 @@ struct DecomposePrintOpConversion : public VectorToSCFPattern<vector::PrintOp> {
     if (vectorType.getRank() != 1) {
       // Flatten n-D vectors to 1D. This is done to allow indexing with a
       // non-constant value (which can currently only be done via
-      // vector.extractelement for 1D vectors).
+      // vector.extract for 1D vectors).
+      // TODO: vector.extract supports N-D non-constant indices now.
```
Comment (banach-space) [inline mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp:760]: Let's either remove this block (preferred) or rephrase the outdated comment - right now it's not clear what the TODO is (missing verb). 

Suggestion (but I'd really prefer this being removed instead):
```cpp
      // Flatten n-D vectors to 1D. This is done for historical reasons and should be removed (using dynamic indices to extract n-D vectors was not supported when this was added).
      // TODO: Remove this block
```

Or something similar :)

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 4]
```diff
@@ -233,10 +233,9 @@ func.func @broadcast_vec2d_from_vec0d(%arg0: vector<f32>) -> vector<3x2xf32> {
 // CHECK-LABEL: @broadcast_vec2d_from_vec0d(
 // CHECK-SAME:  %[[A:.*]]: vector<f32>)
 //       CHECK: %[[T0:.*]] = builtin.unrealized_conversion_cast %[[A]] : vector<f32> to vector<1xf32>
+//       CHECK: %[[T5:.*]] = builtin.unrealized_conversion_cast %[[T0]] : vector<1xf32> to f32
```
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:236]: Hm, why wouldn't this be lowered to `llvm.extractelement`? And is `T5` used at all? In fact, would you mind sharing the output?

**Review by banach-space — COMMENTED** [review]
Thank you so much for the clean-up!

There's a lot of subtle changes here. Would you mind splitting into smaller PRs? For example:
* vectorization
* VectorOps
* etc

?

File: mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp [hunk 5]
```diff
@@ -1614,7 +1613,7 @@ struct Strategy1d<TransferWriteOp> {
 /// This pattern generates IR as follows:
 ///
 /// 1. Generate a for loop iterating over each vector element.
-/// 2. Inside the loop, generate a InsertElementOp or ExtractElementOp,
+/// 2. Inside the loop, generate a InsertOp or ExtractOp,
```
Comment (dcaballe) [inline mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp:1616]: nit: an

**Review by dcaballe — APPROVED** [review]
Thank you so much! LGTM % comments from other reviewers

File: mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp (hunk 3 above)
Comment (MacDue) [reply mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp:760]: Fully dynamic n-D vector.inserts/extracts still do not have a lowering.

**Comment by Groverkss** [comment]
> Thank you so much for the clean-up!
> 
> There's a lot of subtle changes here. Would you mind splitting into smaller PRs? For example:
> 
> * vectorization
> * VectorOps
> * etc
> 
> ?

I'm not sure if i can split the entire patch into a pr for each file, but i can help splitting it up into easier review pieces. I sent out the first split for trivial changes: https://github.com/llvm/llvm-project/pull/116053

**Comment by Groverkss** [comment]
Splitting this patch into smaller pieces

**Pull request closed without merging**